File: app/src/main/java/com/explorify/xplore/xplore/GroupActivity.java
package com.explorify.xplore.xplore;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.os.Build;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;

import jp.wasabeef.picasso.transformations.RoundedCornersTransformation;

/**
 * Created by nikao on 2/12/2017.
 */

public class GroupActivity extends Activity {

/*    private String USERBASE_KEY;
    private String USERBASE_APPID;
    private String USERBASE_URL;*/

    private String leader_id, leader_image_url, group_id, reserveName, startDate, endDate;
    private int reserveID, selectedMemberPos;
    private ImageView leader_image, group_experience_image;
    private Button reserveButton; //TODO CHANGE TO ImageView
    private RelativeLayout memberLayout;
    private TextView leader_lname_text, leader_fname_text, leader_rep_text, leader_age_text, leader_tel_text;
    private TextView member_lname_text, member_fname_text, member_age_text, member_tel_text;
    private TextView startDate_text, endDate_text, groupPrefs_text, extraInfo_text;
    private ProgressBar progressBar;
    private RecyclerView memberRecList;
    private long memberCount;
    private Resources resources;
    //private View divider;
    //private float dividerMoveY;
    private ArrayList<User> members = new ArrayList<>();
/*
    FirebaseApp userBaseApp;
    FirebaseOptions userBaseOptions;
    FirebaseDatabase userDB;*/
    DatabaseReference DBref = FirebaseDatabase.getInstance().getReference();
    Group tempGroup;
    User tempMember;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.group_layout);

        Authorize();
        //buildUserBase();
        PreLoadData();

        resources = getResources();

        //Receiving data from last intent
        Intent intent = this.getIntent();
        group_id = intent.getStringExtra("group_id");
        leader_id = intent.getStringExtra("leader_id");
        leader_image_url = intent.getStringExtra("leader_image_url");
        reserveID = intent.getIntExtra("reserve_id",0);
        reserveName = intent.getStringExtra("reserve_name");

        InitLayout();
        InitMemberList();

        //reset the layout TODO remove after finishing testing and clear layouts
        ResetLayout();

        //Loading the info
        LoadGroupData(group_id);
        ApplyReserveData();
    }

    private void InitLayout()
    {
        //ProgressBar
        progressBar = (ProgressBar) findViewById(R.id.groupProgressBar);
        progressBar.setVisibility(View.VISIBLE);

        //Leader Stuff
        leader_image = (ImageView) findViewById(R.id.leader_profile_image);
        leader_fname_text = (TextView) findViewById(R.id.leader_fname_text);
        leader_lname_text = (TextView) findViewById(R.id.leader_lname_text);
        leader_rep_text = (TextView) findViewById(R.id.leader_rep_text);
        leader_age_text = (TextView) findViewById(R.id.leader_age_text);
        leader_tel_text = (TextView) findViewById(R.id.leader_tel_text);

        //Member Stuff
        member_fname_text = (TextView) findViewById(R.id.member_fname_text);
        member_lname_text = (TextView) findViewById(R.id.member_lname_text);
        member_age_text = (TextView) findViewById(R.id.member_age_text);
        member_tel_text = (TextView) findViewById(R.id.member_tel_text);
        memberLayout = (RelativeLayout) findViewById(R.id.member_profile_layout);
        memberLayout.setVisibility(View.GONE);

        //Group Stuff
        group_experience_image = (ImageView) findViewById(R.id.group_experience_image);
        startDate_text = (TextView) findViewById(R.id.date_start);
        endDate_text = (TextView) findViewById(R.id.date_end);
        groupPrefs_text = (TextView) findViewById(R.id.group_prefs_text);
        extraInfo_text = (TextView) findViewById(R.id.group_extrainfo_text);

        //ReserveButon Stuff
        reserveButton = (Button) findViewById(R.id.group_reserveButton);
        reserveButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                General.openReserveInfoFragment(reserveID, GroupActivity.this);
            }
        });

    }

    private void InitMemberList()
    {
        LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
        memberRecList = (RecyclerView) findViewById(R.id.member_list); //NOTE: uncomment 'member_list' in group_layout.xml
        memberRecList.setHasFixedSize(true);
        memberRecList.setLayoutManager(layoutManager);
    }

    private void ResetLayout()
    {
        //leader
        leader_fname_text.setText("-");
        leader_lname_text.setText("-");
        leader_age_text.setText("");
        leader_tel_text.setText("");
        leader_rep_text.setText("");
        //member
        member_fname_text.setText("-");
        member_lname_text.setText("-");
        member_age_text.setText("");
        member_tel_text.setText("");
        //group
        startDate_text.setText("");
        endDate_text.setText("");
        groupPrefs_text.setText("");
        extraInfo_text.setText("");
    }

    private void Authorize()
    {
        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.default_web_client_id))
                .requestEmail()
                .build();
    }

/*
    private void buildUserBase()
    {
        USERBASE_KEY = getResources().getString(R.string.user_firebase_key);
        USERBASE_APPID = getResources().getString(R.string.firebase_appid);
        USERBASE_URL = getResources().getString(R.string.user_firebase_url);

        userBaseOptions = new FirebaseOptions.Builder()
                .setApiKey(USERBASE_KEY)
                .setApplicationId(USERBASE_APPID)
                .setDatabaseUrl(USERBASE_URL)
                .build();
        try {
            if (FirebaseApp.getApps(this).get(1).equals(null)) {
                FirebaseApp.initializeApp(this, userBaseOptions, "userbase");
            }
        }
        catch (IndexOutOfBoundsException e)
        {
            FirebaseApp.initializeApp(this, userBaseOptions, "userbase");
        }
        userBaseApp = FirebaseApp.getInstance("userbase");
        userDB = FirebaseDatabase.getInstance(userBaseApp);
    }
*/

    private void PreLoadData()
    {
        memberCount = 1;
        selectedMemberPos = -1;
        members.clear();
    }

    private void LoadGroupData(final String groupId)
    {
        DatabaseReference Ref = DBref.child("groups").getRef();
        Query query = Ref.child(groupId);
        query.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {

                //creating the temporary (current) group
                tempGroup = new Group();
                tempGroup.leader = new User();
                tempGroup = dataSnapshot.getValue(Group.class);
                tempGroup.setGroup_id(dataSnapshot.getKey());

                memberCount = tempGroup.getMember_ids().size();

                for(String memberId : tempGroup.getMember_ids()){ //TODO? this returns members in random order
                    GetMemberInfo(String.valueOf(memberId));
                }
            }

            @Override
            public void onCancelled(DatabaseError databaseError) { }
        });
    }

    private void GetMemberInfo(final String userId)
    {
        DatabaseReference ref = DBref.child("users").getRef();
        Query query = ref.child(userId);
        query.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                tempMember = dataSnapshot.getValue(User.class); //Getting member info
                members.add(tempMember); //Setting member info
                memberCount --; //Iterating member index
                if(memberCount == 0) { //Checking if member list retrieval finished
                    ApplyGroupData();
                }

            }

            @Override
            public void onCancelled(DatabaseError databaseError) { }
        });
    }

    private void ApplyReserveData()
    {
        //TODO change this so it doesn't open database again and again
        Context context = GroupActivity.this;
        //TODO convert this to java and skip the other crap arguments
        DBManager dbManager = new DBManager(context, "reserveDB.db", General.DB_TABLE);
        dbManager.openDataBase();

        reserveButton.setText(reserveName);
        reserveButton.setBackgroundResource(dbManager.getImageId(
                reserveID, this, dbManager.getGENERAL_TABLE())
        );
    }

    private void ApplyGroupData()
    {
        progressBar.setVisibility(View.GONE);
        Picasso.with(this)
                .load(leader_image_url)
                .transform(new RoundedCornersTransformation(
                        getResources().getInteger(R.integer.pic_big_angle),
                        getResources().getInteger(R.integer.pic_big_margin)))
                .into(leader_image);
        leader_fname_text.setText(members.get(0).getFname());
        leader_lname_text.setText(members.get(0).getLname());
        leader_age_text.setText(getString(R.string.age) + ": " +
                General.calculateAge(TimeManager.Companion.getGlobalTimeStamp(), members.get(0).getBirth_date()));
        leader_tel_text.setText(getString(R.string.tel)+": "+members.get(0).getTel_num());
        leader_rep_text.setText(String.valueOf(members.get(0).getReputation()));

        if(tempGroup.isExperienced()) {

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
                group_experience_image.setImageDrawable(getResources().getDrawable(R.drawable.ic_check, this.getTheme()));
            else
                group_experience_image.setImageDrawable(getResources().getDrawable(R.drawable.ic_check));
        }
        else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
                group_experience_image.setImageDrawable(getResources().getDrawable(R.drawable.ic_x, this.getTheme()));
            else
                group_experience_image.setImageDrawable(getResources().getDrawable(R.drawable.ic_x));
        }

        group_experience_image.setOnClickListener(new View.OnClickListener() {
              @Override
              public void onClick(View v) {
                  AlertDialog.Builder builder = new AlertDialog.Builder(GroupActivity.this);
                  builder.setTitle(R.string.what_is_this)
                          .setMessage(R.string.group_exp_help)
                          .setPositiveButton(R.string.okay, null);
                  builder.show();
              }
        });

        startDate = BufferDate(tempGroup.getStart_date());
        endDate = BufferDate(tempGroup.getEnd_date());

        startDate_text.setText(startDate);
        endDate_text.setText(endDate);
        groupPrefs_text.setText(tempGroup.getGroup_preferences());
        extraInfo_text.setText(tempGroup.getExtra_info());

        populateMemberImageList();
    }

    private void populateMemberImageList() {
        MemberListAdapter adapter = new MemberListAdapter(GroupActivity.this, members, memberLayout);
        memberRecList.setAdapter(adapter);
    }
    
    //adds slashes to a date given in int (yyyy.mm.dd) without dots
    private String BufferDate(long date)
    {
        StringBuffer sd = new StringBuffer(String.valueOf(date));
        sd.insert(4,"/");
        sd.insert(7,"/");
        return String.valueOf(sd);
    }
}

File: app/src/main/java/com/xplore/GroupActivity.kt
package com.xplore

import android.app.Activity
import android.app.AlertDialog
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.squareup.picasso.Picasso

import java.util.ArrayList

import jp.wasabeef.picasso.transformations.RoundedCornersTransformation
import kotlinx.android.synthetic.main.group_info.*
import kotlinx.android.synthetic.main.reserve_list_item.*

/**
 * Created by nikao on 2/12/2017.
 */

class GroupActivity : Activity() {

    private var group_id: String = ""
    private var reserveID: Int = 0
    private var memberCount: Int = 0

    private val members = ArrayList<User>()
    internal val DBref = FirebaseDatabase.getInstance().reference
    internal val groupsDBref = DBref.child("groups").ref //TODO remove .ref
    internal val usersDBref = DBref.child("users").ref
    internal var tempGroup = Group();
    internal var tempMember = User();

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.group_info)
        //buildUserBase();

        resetVariables()
        receiveData()
        initMemberList()

        //reset the layout TODO remove after finishing testing and clear layouts
        ResetLayout()

        //Loading the info
        loadGroupData(group_id)
        applyReserveData()
    }

    //Resets variables so a new group can be loaded from zero
    private fun resetVariables() {
        memberCount = 1
        members.clear()
    }

    //Receives group data from last intent
    private fun receiveData() {
        val intent = this.intent
        group_id = intent.getStringExtra("group_id")
        reserveID = intent.getIntExtra("reserve_id", 0)
    }

    private fun initMemberList() {
        val layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        membersRecyclerView.setHasFixedSize(true)
        membersRecyclerView.layoutManager = layoutManager
    }

    private fun ResetLayout() {
        //leader
        leaderFnameTextView.text = "-"
        leaderLnameTextView.text = "-"
        leaderAgeTextView.text = ""
        leaderTelTextView.text = ""
        leaderRepTextView.text = ""
        //member
        selectedMemberFnameTextView.text = "-"
        selectedMemberLnameTextView.text = "-"
        selectedMemberAgeTextView.text = ""
        selectedMemberTelTextView.text = ""
        //group
        startDateTextView.text = ""
        endDateTextView.text = ""
        groupPrefsTextView.text = ""
        groupExtraInfoTextView.text = ""
    }

    /*
    //IN CASE WE'RE USING A SECOND FIREBASE DATABASE FOR USERS
    private void buildUserBase()
    {
    USERBASE_KEY = getResources().getString(R.string.user_firebase_key);
    USERBASE_APPID = getResources().getString(R.string.firebase_appid);
    USERBASE_URL = getResources().getString(R.string.user_firebase_url);
    userBaseOptions = new FirebaseOptions.Builder()
    .setApiKey(USERBASE_KEY)
    .setApplicationId(USERBASE_APPID)
    .setDatabaseUrl(USERBASE_URL)
    .build();
    try {
    if (FirebaseApp.getApps(getActivity()).get(1).equals(null)) {
    FirebaseApp.initializeApp(getActivity(), userBaseOptions, "userbase");
    } catch (IndexOutOfBoundsException e) {
    FirebaseApp.initializeApp(getActivity(), userBaseOptions, "userbase");
    }
    userBaseApp = FirebaseApp.getInstance("userbase");
    userDB = FirebaseDatabase.getInstance(userBaseApp);
    }
    */

    private fun loadGroupData(groupId: String) {
        val query = groupsDBref.child(groupId)
        query.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {

                //creating the temporary (current) group
                tempGroup = Group()
                tempGroup = dataSnapshot.getValue(Group::class.java)
                tempGroup.setGroup_id(dataSnapshot.key)

                memberCount = tempGroup.getMember_ids().size.toInt()

                for (memberId in tempGroup.getMember_ids()) { //TODO? this returns members in random order
                    getUserInfo(memberId.toString())
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    //Gets user info from Firebase using userId
    private fun getUserInfo(userId: String) {
        val query = usersDBref.child(userId)
        query.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                tempMember = dataSnapshot.getValue(User::class.java) //Getting member info
                members.add(tempMember) //Setting member info
                memberCount-- //Iterating member index
                if (memberCount == 0) { //Checking if member list retrieval finished
                    ApplyGroupData()
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    private fun applyReserveData() {
        val context = this@GroupActivity //TODO insert this into constructor
        val dbManager = DBManager(context)
        dbManager.openDataBase()
        val tempReserveCard = dbManager.getReserveCard(reserveID)


        reserveCardView.setOnClickListener { General.openReserveInfoFragment(reserveID, context) }
        reserveNameTextView.text = tempReserveCard.name
        reserveImageView.setImageResource(tempReserveCard.imageId)
        reserveIconImageView.setImageResource(ReserveIcons.grey[tempReserveCard.iconId])
    }

    private fun ApplyGroupData() {
        groupProgressBar.visibility = View.GONE
        Picasso.with(this)
                .load(members[0].getProfile_picture_url())
                .transform(RoundedCornersTransformation(
                        resources.getInteger(R.integer.pic_big_angle),
                        resources.getInteger(R.integer.pic_big_margin)))
                .into(leaderImageView)
        leaderFnameTextView.text = members[0].getFname()
        leaderLnameTextView.text = members[0].getLname()
        val age = General.calculateAge(TimeManager.globalTimeStamp, members[0].getBirth_date())
        leaderAgeTextView.text = "${getString(R.string.age)}: $age"
        leaderTelTextView.text = "${getString(R.string.tel)}: ${members[0].getTel_num()}"
        leaderRepTextView.text = members[0].getReputation().toString()

        //Setting experienced icon
        if (tempGroup.isExperienced) {
            groupExpImageView.setImageResource(R.drawable.ic_check)
        } else {
            groupExpImageView.setImageResource(R.drawable.ic_x)
        }

        groupExpImageView.setOnClickListener {
            popExperienceInfoDialog()
        }

        startDateTextView.text = BufferDate(tempGroup.getStart_date())
        endDateTextView.text = BufferDate(tempGroup.getEnd_date())

        groupPrefsTextView.text = tempGroup.getGroup_preferences()
        groupExtraInfoTextView.text = tempGroup.getExtra_info()

        populateMemberImageList()
    }

    //Displays information about the experience icon (X and tick)
    private fun popExperienceInfoDialog() {
        val builder = AlertDialog.Builder(this@GroupActivity)
        builder.setTitle(R.string.what_is_this)
                .setMessage(R.string.group_exp_help)
                .setPositiveButton(R.string.okay, null)
        builder.show()
    }

    private fun populateMemberImageList() {
        val adapter = MemberListAdapter(this@GroupActivity, members, selectedMemberProfileLayout)
        membersRecyclerView.adapter = adapter
    }

    //adds slashes to a date given in int (yyyy.mm.dd) without dots
    private fun BufferDate(date: Long): String {
        val sd = StringBuffer(date.toString())
        sd.insert(4, "/")
        sd.insert(7, "/")
        return sd.toString()
    }
}
